### PR TITLE
feat: add probabilistic calibration metrics and visualization pipeline

### LINKED ISSUE #407: Fix `compute_brier_score` monotonicity bug in CDF interpolation
**Fix `compute_brier_score` monotonicity bug in CDF interpolation**

**Problem**
`compute_brier_score` calls `np.interp(threshold, x_vals, q_arr)` where `x_vals = quantile_forecasts[:, t]` — the BG values at each quantile for timestep `t`. `np.interp` requires its `xp` argument to be non-decreasing, but crossing quantiles (e.g. from TimesFM or Chronos2 where quantile trajectories can cross) silently violate this, producing wrong CDF interpolation and therefore a wrong Brier score with no error raised.

**Fix**
Before calling `np.interp`, sort `x_vals` and `q_arr` together by `x_vals` so the CDF is always presented in non-decreasing order:
 ```python
sort_idx = np.argsort(x_vals)
p_hat[t] = np.interp(threshold, x_vals[sort_idx], q_arr[sort_idx], left=q_arr[0], right=q_arr[-1])
 ```

**Affected models**: TimesFM (full_predictions quantile head can produce crossing quantiles), Chronos2.

### LINKED ISSUE #412: Create Probabilistic Calibration Plots
```python
def compute_pit_values(quantile_forecasts, actuals, quantile_levels):
    """Compute PIT values from quantile forecasts.
    
    Args:
        quantile_forecasts: (n_quantiles, forecast_length)
        actuals: (forecast_length,)
        quantile_levels: list of quantile levels
    Returns:
        (forecast_length,) array of PIT values in [0, 1]
    """
```
PIT Shape:

    - Flat: Well-calibrated
    - U-shaped: Under-dispersed (intervals too narrow)
    - Hump-in-middle: Over-dispersed (intervals too wide)
    - Left-skewed: Model systematically over-predicts
    - Right-skewed: Model systematically under-predicts

We need a script that produces Reliability Diagrams, conditional calibration of quantiles. For each quantile q, compute fraction of actual below predict q-quantile; plot vs. identity line. This means we need the antiques recorded (same as for PIT):
    `ep_result["q_forecast"] = q_forecast.tolist()  # (n_quantiles, forecast_length)`

### PR DESCRIPTION
**Pull request by RobotPsychologist**
## Summary

Adds a full probabilistic calibration evaluation and visualization pipeline for nocturnal hypoglycaemia forecasting. Includes new calibration metrics, six visualization scripts, and minor output path housekeeping.

---

## New Probabilistic Metrics

**probabilistic.py**

Three new functions for evaluating probabilistic forecast calibration:

- `compute_pit_values(q_forecasts_batch, actuals_batch, quantile_levels)` — Probability Integral Transform values; uniform distribution under a perfectly calibrated model
- `compute_reliability_curve(q_forecasts_batch, actuals_batch, quantile_levels)` — empirical vs. nominal coverage per quantile level
- `compute_ece(q_forecasts_batch, actuals_batch, quantile_levels)` — Expected Calibration Error (scalar summary of reliability)

All three operate on the `(n_episodes, n_quantiles, horizon)` quantile forecast format already saved in `forecasts.npz`.

**test_probabilistic_metrics.py** — 20 new tests covering edge cases, shape contracts, and numerical correctness (286 tests total passing).

---

## New Visualization Scripts

All scripts read from nocturnal_forecasting and write to results.

### plot_pit_histograms.py
PIT histograms for assessing calibration. Produces:
- **Main body figure**: 2×3 grid (one panel per model), pooled across datasets
- **Appendix figure**: per-model × per-dataset breakdown

### plot_reliability_diagrams.py
Reliability (calibration) diagrams. One panel per model showing empirical vs. nominal coverage curves with ECE annotated in the legend.

### plot_coverage_error_2d.py
2D heatmap of coverage error (empirical − nominal) across quantile levels and forecast horizons, faceted by model.

### plot_coverage_error_2d_sharpness.py
Same layout as above but overlays sharpness (prediction interval width) to distinguish over-confidence from under-confidence.

### plot_coverage_sharpness_scatter.py
Scatter plot: mean coverage error vs. sharpness, one point per model×dataset combination. Defaults to best run per model×dataset (`best_by_model_dataset.csv`); `--all-runs` flag exposes all checkpoints.

### plot_chronos2_forecast_grid.py
4×5 episode grid for the Chronos-2 model. Rows = datasets (`aleppo_2017`, `brown_2019`, `lynch_2022`, `tamborlane_2008`); columns = P10/P25/P50/P75/P90 RMSE percentile episodes selected from `forecasts.npz`. Each panel shows the full quantile fan, actual BG trace, and annotates episode ID, RMSE, and minimum BG. Y-axis fixed 0–22 mmol/L across all panels.

---

## Housekeeping

- results added to .gitignore — generated figures are not tracked
- compare_forecasts.py and visualize_chronos2_episodes.py: output directories redirected from `images/figures/` → results to consolidate all generated artefacts in one location

---

## Testing

```
pytest tests/evaluation/test_probabilistic_metrics.py -v   # 20 new tests
pytest                                                       # 286 total, all passing
```

### CODE REVIEW

File: scripts/visualization/plot_reliability_diagrams.py [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+            else:
+                weighted_emp += empirical * n
+        assert nominal_ref is not None and weighted_emp is not None
+        agg_empirical = weighted_emp / total_n
+
+        # ECE from the aggregated curve
+        from scipy.integrate import trapezoid
+
+        agg_ece = float(trapezoid(np.abs(agg_empirical - nominal_ref), nominal_ref))
+
```
[the remark this answers is not in the document]
Comment (RobotPsychologist) [reply scripts/visualization/plot_reliability_diagrams.py:261]: 1. Replaced the from scipy.integrate import trapezoid import (which was inside the plotting loop) with np.trapz(...), removing the extraneous scipy dependency for this operation.
2. Added an assert np.allclose(nominal, nominal_ref) before accumulating empirical curves, so incompatible quantile grids across datasets will raise an explicit error rather than silently producing a meaningless average.

File: scripts/visualization/plot_pit_horizon_heatmap.py [hunk 2]
```diff
@@ -0,0 +1,659 @@
...
+        step_bins: Number of time-step groups on the X-axis.  Must divide
+            forecast_length evenly, or the last group is dropped.
+
+    Returns:
+        density: Shape (n_bins, step_bins).  Each column is a normalised
+            histogram of PIT values for that group of steps.  Uniform
+            calibration → all values ≈ 1.0.
+        bin_edges: Shape (n_bins + 1,) — edges of the PIT-value bins.
+        mad_per_bin: List of mean absolute density deviations from 1.0 per time bin.
+    """
```
[the remark this answers is not in the document]
Comment (RobotPsychologist) [reply scripts/visualization/plot_pit_horizon_heatmap.py:290]: You're right — the [List[Tuple[float, float]]](http://vscodecontentref/3) return annotation was a leftover from when the function returned KS (statistic, p_value) pairs, before that was replaced with MAD. Fixed in the latest commit: annotation is now [List[float]](vscode-file://vscode-app/snap/code/232/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) and the docstring opening line updated to "per-bin MAD values".

File: scripts/visualization/plot_coverage_error_2d_sharpness.py [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+Y-axis : 0.80 − coverage_80  (80% PI coverage error)
+3rd dim: sharpness (three alternative encodings — one panel each)
+
+Panel A — Marker size proportional to mean normalised sharpness.
+           Larger = wider / less-sharp intervals.
+
+Panel B — Marker colour saturation (alpha) encodes sharpness, with model
+           identity carried by marker shape only.  Desaturated = sharper;
+           opaque = wider.  This sacrifices the model-colour encoding in
+           exchange for a continuous sharpness scale.
```
[the remark this answers is not in the document]
Comment (RobotPsychologist) [reply scripts/visualization/plot_coverage_error_2d_sharpness.py]: the docstring actively misdescribes the encoding, which would confuse anyone reading the output figure. Panel B actually uses [MODEL_COLORS](vscode-file://vscode-app/snap/code/232/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) for colour and DATASET_MARKERS for shape (identical to Panel A), with alpha substituted for size as the sharpness dimension. The claim that model-colour encoding was "sacrificed" is simply wrong. Fixed the module docstring to match the implementation

**Comment by RobotPsychologist** [comment]
Addressing the three remaining reviewer comments:

**1. `_validate_batch_quantile_inputs` missing monotonicity check** (`src/evaluation/metrics/probabilistic.py`)
Added a three-tier check on `np.diff(quantile_forecasts_batch, axis=1)`:
- ≤ 0.01 mmol/L → `warnings.warn` (floating-point noise)
- 0.01–1.0 mmol/L → `np.sort(axis=1)` in-place + `warnings.warn` (recoverable crossing, e.g. normal TimesFM quantile head behaviour)
- > 1.0 mmol/L → `raise ValueError` (catastrophic, e.g. confirmed high-LR TimesFM bad runs with sharpness_50 ≈ −742 mmol/L)

**2. `compute_pit_values` docstring — truncated quantile grid makes Uniform[0,1] claim misleading** (`src/evaluation/metrics/probabilistic.py`)
Replaced "PIT values follow Uniform[0, 1]" with a note clarifying that with a truncated grid (e.g. 0.1–0.9), hard clamping produces point masses at 0 and 1 even under perfect calibration; the distribution is only uniform within (q_min, q_max). The histogram should be interpreted relative to the quantile range, not against a flat Uniform[0,1] baseline.

**3. Dead no-op loop in `plot_main_body`** (`scripts/visualization/plot_reliability_diagrams.py`)
Removed the empty `for …: pass` loop. It was a leftover placeholder from an earlier drafting pass — the actual weighted-average aggregation was already implemented directly below it.

**Comment by RobotPsychologist** [comment]
Addressing all 4 new reviewer comments:

**1. `compute_brier_score` CDF non-monotone after sort** (`PRRT_kwDONflgos59tWb9`)
You're right — `q_arr[sort_idx]` reindexed probabilities by BG-sort order, meaning when quantiles cross (e.g. q90 BG < q50 BG), the interpolation table had decreasing probabilities as BG increased: an invalid CDF. The fix is `np.sort(x_vals)` paired with `q_arr` unchanged, which is exactly what `_validate_batch_quantile_inputs`'s `np.sort(axis=1)` does in the 3D path.

**2. Minor inversions (≤ 0.01 mmol/L) pass through unsorted** (`PRRT_kwDONflgos59tWcG`)
Agreed — `compute_pit_values`'s bracket-finding uses `hi = (xp <= y).sum(axis=1)` which assumes sorted rows, so even a sub-0.01 mmol/L crossing can assign the wrong bracket (the `denom ≤ 0` fallback limits the damage but doesn't fix the root cause). Now sorts at every inversion tier; the warning tiers are preserved but the unsorted-pass-through path is removed.

**3. `np.asarray` passes the caller's array through; in-place sort mutates it** (`PRRT_kwDONflgos59tWcM`)
Exactly right. Changed to `np.array(..., dtype=np.float64)` (always copies) in `compute_pit_values`, `compute_reliability_curve`, `compute_coverage`, and `compute_sharpness_by_step`. `test_moderate_quantile_inversions_sorted` is updated to verify the caller's array is *not* mutated.

**4. `compute_ece` docstring "normalised to [0, 1]"** (`PRRT_kwDONflgos59tWcO`)
Removed — the `Returns` block already correctly states `ECE in [0, 0.5]`. The phrase was from an earlier draft when I was conflating ECE with a percentage.

**Comment by RobotPsychologist** [comment]
Good catch — the reviewer is correct. `y == xp.max()` was landing in the `above` branch (`hi == n_quantiles`) and getting clamped to PIT = 1.0, when the correct value is `q_arr[-1]` (e.g. 0.9). The lower boundary was already handled correctly (y == xp.min() falls into `inside` and interpolates to `q_arr[0]`), so this was an asymmetry.

Fix: changed `above` to use `(xp < y).all(axis=1)` (strictly greater than max) and clamped `hi_idx` to `n_quantiles - 1` in the bracket lookup so the interpolation for `y == xp.max()` correctly produces `weight = 1.0 → q_arr[-1]`.

In practice this has no visible effect on continuous BG data (exact floating-point ties with a predicted quantile value are essentially measure-zero), but the code is now correct for discrete or rounded inputs too.

**Review by RobotPsychologist — COMMENTED** [review]
Good work!
